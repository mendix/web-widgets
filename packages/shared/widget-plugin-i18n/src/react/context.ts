import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
