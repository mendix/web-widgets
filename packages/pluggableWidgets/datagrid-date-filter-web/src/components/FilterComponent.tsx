import { FilterSelector } from "@mendix/widget-plugin-filter-selector/FilterSelector";
import { useWatchValues } from "@mendix/widget-plugin-hooks/useWatchValues";
import { createElement, CSSProperties, ReactElement, useCallback, useRef, useState } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";
import classNames from "classnames";
import DatePickerComponent from "react-datepicker";
import { useSetInitialConditionEffect } from "../features/initialize";
import { DatePicker, RangeDateValue } from "./DatePicker";
import { useOnResetValueEvent } from "@mendix/widget-plugin-external-events/hooks";

export interface FilterComponentProps {
    adjustable: boolean;
    calendarStartDay?: number;
    className?: string;
    defaultFilter: DefaultFilterEnum;
    defaultValue?: Date;
    defaultStartDate?: Date;
    defaultEndDate?: Date;
    dateFormat?: string;
    locale?: string;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (value: Date | undefined, rangeValues: RangeDateValue, type: DefaultFilterEnum) => void;
    name: string;
    parentChannelName: string | null;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Date | undefined>(props.defaultValue);
    const [rangeValues, setRangeValues] = useState<RangeDateValue>([props.defaultStartDate, props.defaultEndDate]);
    const pickerRef = useRef<DatePickerComponent | null>(null);
    const reset = useCallback(() => {
        setValue(undefined);
        setRangeValues([undefined, undefined]);
    }, []);

    useOnResetValueEvent({
        widgetName: props.name,
        parentChannelName: props.parentChannelName ?? undefined,
        listener: reset
    });

    useWatchValues(
        (_prev, _next) => {
            props.updateFilters?.(value, rangeValues, type);
        },
        [value, rangeValues, type]
    );

    const focusInput = useCallback(() => {
        if (pickerRef.current) {
            pickerRef.current.setFocus();
        }
    }, [pickerRef.current]);

    useSetInitialConditionEffect({
        initialFilterType: type,
        initialFilterValue: value ?? rangeValues,
        updateFilters: props.updateFilters
    });

    return (
        <div
            className={classNames("filter-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            style={props.styles}
        >
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    defaultFilter={props.defaultFilter}
                    id={props.id}
                    onChange={useCallback(
                        type => {
                            setType(prev => {
                                if (prev === type) {
                                    return prev;
                                }
                                focusInput();
                                return type;
                            });
                        },
                        [focusInput]
                    )}
                    options={
                        [
                            { value: "between", label: "Between" },
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" },
                            { value: "empty", label: "Empty" },
                            { value: "notEmpty", label: "Not empty" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                calendarStartDay={props.calendarStartDay}
                dateFormat={props.dateFormat}
                disabledInput={type === "empty" || type === "notEmpty"}
                enableRange={type === "between"}
                locale={props.locale}
                id={props.id}
                placeholder={props.placeholder}
                rangeValues={rangeValues}
                ref={pickerRef}
                screenReaderCalendarCaption={props.screenReaderCalendarCaption}
                screenReaderInputCaption={props.screenReaderInputCaption}
                setRangeValues={setRangeValues}
                setValue={setValue}
                value={value}
            />
        </div>
    );
}
