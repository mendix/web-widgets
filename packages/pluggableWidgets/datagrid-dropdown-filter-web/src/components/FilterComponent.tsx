import { useWatchValues } from "@mendix/widget-plugin-hooks/useWatchValues";
import { createElement, CSSProperties, ReactElement, UIEventHandler, useCallback, useEffect } from "react";
import { useSelectState } from "../features/select";
import { EMPTY_OPTION_VALUE, finalizeOptions, parseInitValues } from "../features/setup";
import { Option } from "../utils/types";
import { SelectComponent } from "./SelectComponent";
import { useOnResetValueEvent } from "@mendix/widget-plugin-external-events/hooks";

export interface FilterComponentProps {
    ariaLabel?: string;
    className?: string;
    initialSelected?: string;
    status?: JSX.Element;
    footer?: JSX.Element;
    emptyOptionCaption?: string;
    multiSelect?: boolean;
    id?: string;
    options: Option[];
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (values: Option[]) => void;
    onTriggerClick?: () => void;
    onContentScroll?: UIEventHandler<HTMLUListElement>;
    name: string;
    parentChannelName: string | null;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const multiSelect = !!props.multiSelect;
    const options = finalizeOptions(props.options, { multiSelect, emptyOptionCaption: props.emptyOptionCaption });
    const [state, { toggle, setSelected, reset }] = useSelectState(
        options,
        parseInitValues(props.initialSelected ?? "")
    );

    const onSelect = useCallback((value: string) => {
        if (multiSelect) {
            toggle(value);
        } else {
            setSelected(value === EMPTY_OPTION_VALUE ? [] : [value]);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useSetInitialConditionEffect(props.updateFilters, options, state.selected);

    useWatchValues(
        (_, [selected]) => {
            props.updateFilters?.(selected.length > 0 ? options.filter(o => selected.includes(o.value)) : []);
        },
        [state.selected]
    );

    useOnResetValueEvent({
        widgetName: props.name,
        parentChannelName: props.parentChannelName ?? undefined,
        listener: reset
    });

    return (
        <SelectComponent
            ariaLabel={props.ariaLabel}
            className={props.className}
            footer={props.footer}
            id={props.id}
            inputValue={state.inputValue}
            multiSelect={multiSelect}
            onSelect={onSelect}
            options={options}
            selected={state.selected}
            status={props.status}
            styles={props.styles}
            tabIndex={props.tabIndex}
            placeholder={props.emptyOptionCaption}
            onContentScroll={props.onContentScroll}
            onTriggerClick={props.onTriggerClick}
        />
    );
}

function useSetInitialConditionEffect(
    updateFilters: ((values: Option[]) => void) | undefined,
    options: Option[],
    selected: string[]
): void {
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(() => updateFilters?.(options.filter(o => selected.includes(o.value))), []);
}
