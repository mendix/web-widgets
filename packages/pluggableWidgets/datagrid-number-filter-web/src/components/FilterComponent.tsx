import {
    BaseProps,
    BigHelper,
    FilterList,
    InputWithFilters,
    useInputProps
} from "@mendix/widget-plugin-filtering/controls";
import { createElement } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridNumberFilterProps";

const filterDefs: Record<DefaultFilterEnum, string> = {
    greater: "Greater than",
    greaterEqual: "Greater than or equal",
    equal: "Equal",
    notEqual: "Not equal",
    smaller: "Smaller than",
    smallerEqual: "Smaller than or equal",
    empty: "Empty",
    notEmpty: "Not empty"
};

const filters: FilterList<DefaultFilterEnum> = Object.entries(filterDefs).map(
    ([value, label]: [DefaultFilterEnum, string]) => ({
        value,
        label
    })
);

export interface FilterComponentProps extends BaseProps {
    defaultFilter: DefaultFilterEnum;
    value: Big | undefined;
    onChange: (value: Big | undefined, type: DefaultFilterEnum) => void;
    parentChannelName: string | null;
    changeDelay?: number;
}

export function FilterComponent(props: FilterComponentProps): React.ReactElement {
    const { defaultFilter, value, onChange, parentChannelName, changeDelay, ...baseProps } = props;
    const inputProps = useInputProps({
        name: baseProps.name,
        parentChannelName,
        inputType: "number",
        inputDisabled: filter => filter === "empty" || filter === "notEmpty",
        changeDelay,
        defaultFilter,
        filters,
        value,
        onChange,
        valueHelper: BigHelper
    });
    return <InputWithFilters {...baseProps} {...inputProps} />;
}
