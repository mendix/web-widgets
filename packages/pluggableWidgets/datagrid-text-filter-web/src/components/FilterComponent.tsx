import {
    BaseProps,
    StringHelper,
    FilterList,
    InputWithFilters,
    useInputProps
} from "@mendix/widget-plugin-filtering/controls";
import { createElement } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridTextFilterProps";

const filterDefs: Record<DefaultFilterEnum, string> = {
    contains: "Contains",
    startsWith: "Starts with",
    endsWith: "Ends with",
    greater: "Greater than",
    greaterEqual: "Greater than or equal",
    equal: "Equal",
    notEqual: "Not equal",
    smaller: "Smaller than",
    smallerEqual: "Smaller than or equal",
    empty: "Empty",
    notEmpty: "Not empty"
};

const filters: FilterList<DefaultFilterEnum> = Object.entries(filterDefs).map(
    ([value, label]: [DefaultFilterEnum, string]) => ({
        value,
        label
    })
);

export interface FilterComponentProps extends BaseProps {
    defaultFilter: DefaultFilterEnum;
    value: string | undefined;
    onChange: (value: string | undefined, type: DefaultFilterEnum) => void;
    parentChannelName: string | null;
    changeDelay?: number;
}

export function FilterComponent(props: FilterComponentProps): React.ReactElement {
    const { defaultFilter, value, onChange, parentChannelName, changeDelay, ...baseProps } = props;
    const inputProps = useInputProps({
        name: baseProps.name,
        parentChannelName,
        inputType: "text",
        inputDisabled: filter => filter === "empty" || filter === "notEmpty",
        changeDelay,
        defaultFilter,
        filters,
        value,
        onChange,
        valueHelper: StringHelper
    });
    return <InputWithFilters {...baseProps} {...inputProps} />;
}
