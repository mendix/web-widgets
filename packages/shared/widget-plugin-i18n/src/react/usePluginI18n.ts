import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
