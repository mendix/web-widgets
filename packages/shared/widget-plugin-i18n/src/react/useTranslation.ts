import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
