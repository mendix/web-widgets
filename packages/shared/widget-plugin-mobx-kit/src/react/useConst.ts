import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
