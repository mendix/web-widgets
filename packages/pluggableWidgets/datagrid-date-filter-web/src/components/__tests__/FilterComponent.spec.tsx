import { render } from "enzyme";
import { createElement } from "react";
import { render as render_fromTestingLibrary } from "@testing-library/react";
import { FilterComponent, FilterComponentProps } from "../FilterComponent";
import ReactDOM from "react-dom";

const commonProps: FilterComponentProps = {
    name: "DateFilter",
    parentChannelName: null,
    adjustable: true,
    defaultFilter: "equal"
};

describe("Filter component", () => {
    beforeAll(() => {
        jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);

        // @ts-ignore
        jest.spyOn(ReactDOM, "createPortal").mockReturnValue((element, node) => {
            return element;
        });
    });

    it("renders correctly", () => {
        const component = render(<FilterComponent {...commonProps} />);

        expect(component).toMatchSnapshot();
    });

    it("renders correctly when not adjustable by user", () => {
        const component = render(<FilterComponent {...commonProps} adjustable={false} />);

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with aria labels", () => {
        const component = render(
            <FilterComponent
                {...commonProps}
                screenReaderButtonCaption="my label"
                screenReaderInputCaption="my label"
            />
        );

        expect(component).toMatchSnapshot();
    });

    describe("with defaultValue", () => {
        it("don't call updateFilters when defaultValue get new value", () => {
            const date = new Date(946684800000);
            const updateFilters = jest.fn();
            const { rerender } = render_fromTestingLibrary(
                <FilterComponent {...commonProps} defaultValue={date} updateFilters={updateFilters} />
            );

            // First time updateFilters is called on initial mount
            expect(updateFilters).toBeCalledTimes(1);
            expect(updateFilters).toHaveBeenLastCalledWith(date, [undefined, undefined], "equal");

            const nextValue = new Date(999999900000);

            rerender(<FilterComponent {...commonProps} defaultValue={nextValue} updateFilters={updateFilters} />);

            expect(updateFilters).toBeCalledTimes(1);
            expect(updateFilters).toHaveBeenLastCalledWith(date, [undefined, undefined], "equal");
        });

        it("don't call updateFilters when defaultValue get same value", () => {
            const date = new Date(946684800000);
            const updateFilters = jest.fn();
            const { rerender } = render_fromTestingLibrary(
                <FilterComponent {...commonProps} defaultValue={date} updateFilters={updateFilters} />
            );

            // First time updateFilters is called on initial mount
            expect(updateFilters).toBeCalledTimes(1);
            expect(updateFilters.mock.calls[0][0]).toBe(date);

            const nextValue = new Date(946684800000);

            rerender(<FilterComponent {...commonProps} defaultValue={nextValue} updateFilters={updateFilters} />);

            expect(updateFilters).toBeCalledTimes(1);
        });
    });
});
