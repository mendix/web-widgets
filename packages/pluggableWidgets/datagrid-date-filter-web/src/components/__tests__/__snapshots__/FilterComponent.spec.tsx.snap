// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders correctly 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <div
    class="filter-selector"
  >
    <div
      class="filter-selector-content"
    >
      <button
        aria-controls="undefined-filter-selectors"
        aria-expanded="false"
        aria-haspopup="true"
        class="btn btn-default filter-selector-button button-icon equal"
      >
         
      </button>
    </div>
  </div>
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />
  <span
    class="sr-only"
    id="undefined-label"
  />
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control filter-input"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`Filter component renders correctly when not adjustable by user 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />
  <span
    class="sr-only"
    id="undefined-label"
  />
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`Filter component renders correctly with aria labels 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <div
    class="filter-selector"
  >
    <div
      class="filter-selector-content"
    >
      <button
        aria-controls="undefined-filter-selectors"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="my label"
        class="btn btn-default filter-selector-button button-icon equal"
      >
         
      </button>
    </div>
  </div>
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />
  <span
    class="sr-only"
    id="undefined-label"
  >
    my label
  </span>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control filter-input"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
